include/mata/sparse-set.hh

```diff
@@ -0,0 +1,355 @@
+                    reserve(val + 1);
+
+                dense[size_] = val;
+                sparse[val] = size_;
+                ++size_;
+            }
+
+            assert(consistent());
+        }
+
+        void erase(const Number val) {
+            if (has(val)) {
+                dense[sparse[val]] = dense[size_ - 1];
+                sparse[dense[size_ - 1]] = sparse[val];
+                --size_;
+            }
+        }
+
+
+
+//////////////////////////////////////////////////////////////////////////////////
+///     New Mata code
+//////////////////////////////////////////////////////////////////////////////////
+
+// Constructors:
+
+        SparseSet(Number size = 0, bool val = false): sparse(size, val), dense(size) {
+            if (val) {
+                size_ = size;
+                for (Number e = 0; e < size; ++e)
+                    dense.push_back(e);
+            }
+            else
+                size_ = 0;
+
+            assert(consistent());
+        };
+
+        //TODO: should this be explicit too? it would mean a lot of fixing expecially in tests, unfortunatelly
+        SparseSet(std::initializer_list<Number> list) {
+            insert(list.begin(),list.end());
+
+            assert(consistent());
+        }
+
+        template <class InputIterator>
+        explicit SparseSet(InputIterator first, InputIterator last)
+        {
+            insert(first,last);
+
+            assert(consistent());
+        }
+
+        template <class T>
+        explicit SparseSet(T & container)
+        {
+            insert(container.begin(),container.end());
+
+            assert(consistent());
+        }
+
+        //This is actually weird, what if the vector is not boolean valaues but values themselves?
+        explicit SparseSet(const std::vector<char> &bv) {
+            reserve(bv.size());
+            for (size_t i = 0; i < bv.size(); i++) {
+                if (bv[i])
+                    insert(i);
+            }
+
+            assert(consistent());
+        }
+
+        SparseSet(const SparseSet<Number>& rhs) = default;
+        SparseSet(SparseSet<Number>&& other) noexcept
+                : sparse{ std::move(other.sparse) }, dense{ std::move(other.dense) },
+                  size_{ other.size_}, capacity_{ other.capacity_} {
+            other.size_ = 0;
+            other.capacity_ = 0;
+
+            assert(consistent());
+        }
+
+        SparseSet<Number>& operator=(const SparseSet <Number>& rhs) = default;
+        SparseSet<Number>& operator=(SparseSet<Number>&& other) noexcept {
+            if (this != &other) {
+                dense = std::move(other.dense);
+                sparse = std::move(other.sparse);
+                size_ = other.size_;
+                capacity_ = other.capacity_;
+                other.size_ = 0;
+                other.capacity_ = 0;
+            }
+
+            assert(consistent());
+
+            return *this;
+        }
+
+// Things
```

Comment:
Not very descriptive. Do we need such comments for sections? It is a matter of taste.